### PR TITLE
Limit the number of IPs addrman learns from each DNS seeder

### PR DESCRIPTION
**Pull request by EthanHeilman**
A risk exists where a malicious DNS seeder eclipses a node by returning an enormous number of IP addresses. In this commit we mitigate this risk by limiting the number of IP addresses addrman learns to 256 per DNS seeder.

As discussed with @theuni

### CODE REVIEW

**Review by randolf — APPROVED** [review]
This is a very good first step in mitigating this type of DoS attack, and 256 seems me to be an extremely generous default.

**Comment by sipa** [comment]
Since DNS responses generally are sent over UDP, all of them need to fit in a single IP packet (I believe), which puts a natural limit regardless. Having some explicit limit sounds good though.

**Comment by randolf** [comment]
@sipa Packets larger than 512 bytes are supported with the introduction of EDNS (RFC 6891 which is dated April 2013, but there are earlier RFCs that also reference EDNS0 that may also be of interest) that uses an unsigned 16-bit integer to specify RDLEN (Record Data Length).  Also, while UDP is a MUST for DNS services, TCP is a SHOULD, and both of these transport layer protocols can, for the most part, support EDNS's larger packet size options.

In summary, the natural limit that is more well-known has effectively been extended (IP packet fragmentation and reassembly make it possible to venture beyond the MSU, which is commonly set to 1,500 bytes).

**Comment by sipa** [comment]
@randolf Thanks, TIL.

**Comment by sdaftuar** [comment]
utACK 46e7f800bd78aa4d4de5915b4a7e5a3234c507d6

**Review by theuni — APPROVED** [review]
utACK 46e7f800bd78aa4d4de5915b4a7e5a3234c507d6

**Comment by EthanHeilman** [comment]
Three years ago I tested the number of DNS entries I could get into Bitcoin for the eclipse attack paper. My test setup was Ubuntu Linux running Bitcoind querying a custom DNS server on localhost. We didn't end up using this attack so I wrote up a blog entry about the general question without mentioning bitcoin: [How many IP addresses can a DNS query return?](http://ethanheilman.tumblr.com/post/110920218915/how-many-ip-addresses-can-a-dns-query-return)

**Comment by fanquake** [comment]
utACK 46e7f80
